openedx/frontend-app-authoring
# Feat  course optimizer page

## Description

Course Optimizer is a [feature approved by the Openedx community](https://github.com/openedx/platform-roadmap/issues/388) that adds a "Course Optimizer" page to studio where users can run a scan of a course for broken links - links that point to pages that have a 404.

Depends on backend: https://github.com/openedx/edx-platform/pull/35887 - test together.

This also requires adding a nav menu item to edx-platform legacy studio. That should be implemented before enabling the waffle flag on prod.

## Links

- [Internal JIRA ticket](https://2u-internal.atlassian.net/browse/TNL-11809)
- [Course Optimizer Discovery](https://2u-internal.atlassian.net/wiki/spaces/TNL/pages/1426587703/TNL-11744+Course+Optimizer+Discovery)
- [Openedx community proposal](https://github.com/openedx/platform-roadmap/issues/388)

## Testing instructions

- Checkout [Backend branch](https://github.com/openedx/edx-platform/pull/35887) as well as this branch
- Add the required waffle flag like this:
- Go to localhost:18000/admin/waffle/flag/
- Click "Add Flag"
- As name, type: "contentstore.enable_course_optimizer"
- In the "everyone" field, select "Yes"
- Leave other fields as they are and save
- Go to a course and insert a broken link to a text block, e.g. just write "https://broken.example.com" somewhere in the block
- Add an image to a text block
- Go to Content -> Files in the menu
- Select that image and lock it
- Go to course-authoring (not legacy studio) in a course and select "Course Optimizer" from Tools menu
- Start a course scan (may take a while)
- Progress steps should appear and show in-progress, then turn green
- Result should show
- Result should list all and only sections with broken or locked links (may be more than the ones you added if there are already broken links in the course)
- "Go to block" links should bring you to the block with the broken or locked link
- Next to it in the results display the broken link should be listed
- You should see the date of the last scan, which should be today
- Refresh page in browser
- Same results as before should load without needing to start a new scan
- Remove any broken and locked links from course
- Refresh page
- Outdated results should still be shown
- Run new scan
- Now, no results should be shown, and the page should signal that no broken links were found
